File: agents/core_agent/src/EventBus.js
/**
 * EventBus - High-performance event system with priority queues and filtering
 */

class EventBus {
  constructor() {
    this.listeners = new Map();
    this.priorities = new Map();
    this.middleware = [];
    this.metrics = {
      eventsEmitted: 0,
      eventsHandled: 0,
    };
  }

  on(event, handler, priority = 5) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, []);
      this.priorities.set(event, []);
    }

    const handlers = this.listeners.get(event);
    const priorities = this.priorities.get(event);

    // Insert based on priority (higher = earlier)
    const insertIndex = priorities.findIndex(p => p < priority);
    const index = insertIndex === -1 ? handlers.length : insertIndex;

    handlers.splice(index, 0, handler);
    priorities.splice(index, 0, priority);

    // Return unsubscribe function
    return () => this.off(event, handler);
  }

  once(event, handler, priority = 5) {
    const onceWrapper = data => {
      this.off(event, onceWrapper);
      handler(data);
    };
    return this.on(event, onceWrapper, priority);
  }

  off(event, handler) {
    const handlers = this.listeners.get(event);
    if (!handlers) return;

    const index = handlers.indexOf(handler);
    if (index > -1) {
      handlers.splice(index, 1);
      this.priorities.get(event).splice(index, 1);
    }
  }

  async emit(event, data) {
    this.metrics.eventsEmitted++;

    // Run middleware
    let processedData = data;
    for (const mw of this.middleware) {
      processedData = await mw(event, processedData);
      if (processedData === null) return; // Cancelled by middleware
    }

    const handlers = this.listeners.get(event) || [];
    const wildcardHandlers = this.listeners.get('*') || [];

    // Execute handlers in priority order
    const allHandlers = [
      ...handlers.map(h => ({ handler: h, type: 'specific' })),
      ...wildcardHandlers.map(h => ({ handler: h, type: 'wildcard' })),
    ];

    for (const { handler } of allHandlers) {
      try {
        await handler(processedData);
        this.metrics.eventsHandled++;
      } catch (error) {
        console.error(`[EventBus] Handler error for ${event}:`, error);
        this.emit('bus:error', { event, error, data: processedData });
      }
    }
  }

  use(middleware) {
    this.middleware.push(middleware);
  }

  getMetrics() {
    return { ...this.metrics };
  }
}

export default EventBus;

File: custom-plugins/cordova-plugin-aicore/cordova-plugin-wifi-billing-agent/src/ios/EventBus.m
//
//  EventBus.m
//  WiFi Billing Agent
//

#import "EventBus.h"

@interface EventHandler : NSObject
@property (nonatomic, copy) void (^handler)(NSDictionary *);
@property (nonatomic, assign) NSInteger priority;
@property (nonatomic, assign) BOOL once;
@end

@implementation EventHandler
@end

@interface EventBus()

@property (nonatomic, strong) NSMutableDictionary<NSString *, NSMutableArray<EventHandler *> *> *listeners;
@property (nonatomic, strong) NSMutableArray<BOOL (^)(NSString *, NSDictionary *)> *middleware;
@property (nonatomic, assign) NSUInteger eventsEmitted;
@property (nonatomic, assign) NSUInteger eventsHandled;

@end

@implementation EventBus

- (instancetype)init {
    self = [super init];
    if (self) {
        _listeners = [[NSMutableDictionary alloc] init];
        _middleware = [[NSMutableArray alloc] init];
        _eventsEmitted = 0;
        _eventsHandled = 0;
    }
    return self;
}

- (void)on:(NSString *)event handler:(void (^)(NSDictionary *))handler {
    [self on:event handler:handler priority:5];
}

- (void)on:(NSString *)event handler:(void (^)(NSDictionary *))handler priority:(NSInteger)priority {
    if (!self.listeners[event]) {
        self.listeners[event] = [[NSMutableArray alloc] init];
    }
    
    EventHandler *wrapper = [[EventHandler alloc] init];
    wrapper.handler = handler;
    wrapper.priority = priority;
    wrapper.once = NO;
    
    // Insert based on priority (higher = earlier)
    NSMutableArray *handlers = self.listeners[event];
    NSUInteger insertIndex = handlers.count;
    for (NSUInteger i = 0; i < handlers.count; i++) {
        if (handlers[i].priority < priority) {
            insertIndex = i;
            break;
        }
    }
    [handlers insertObject:wrapper atIndex:insertIndex];
}

- (void)once:(NSString *)event handler:(void (^)(NSDictionary *))handler {
    if (!self.listeners[event]) {
        self.listeners[event] = [[NSMutableArray alloc] init];
    }
    
    EventHandler *wrapper = [[EventHandler alloc] init];
    wrapper.handler = handler;
    wrapper.priority = 5;
    wrapper.once = YES;
    
    [self.listeners[event] addObject:wrapper];
}

- (void)off:(NSString *)event handler:(void (^)(NSDictionary *))handler {
    NSMutableArray *handlers = self.listeners[event];
    if (!handlers) return;
    
    NSMutableArray *toRemove = [NSMutableArray array];
    for (EventHandler *wrapper in handlers) {
        // Compare handlers by pointer
        if (wrapper.handler == handler) {
            [toRemove addObject:wrapper];
        }
    }
    [handlers removeObjectsInArray:toRemove];
}

- (void)emit:(NSString *)event data:(NSDictionary *)data {
    self.eventsEmitted++;
    
    // Run middleware
    __block NSDictionary *processedData = data;
    for (BOOL (^mw)(NSString *, NSDictionary *) in self.middleware) {
        if (!mw(event, processedData)) {
            return; // Cancelled
        }
    }
    
    // Get handlers for this event and wildcard
    NSMutableArray *handlers = [self.listeners[event] mutableCopy] ?: [NSMutableArray array];
    NSArray *wildcardHandlers = self.listeners[@"*"] ?: @[];
    [handlers addObjectsFromArray:wildcardHandlers];
    
    // Execute handlers
    NSMutableArray *onceHandlers = [NSMutableArray array];
    
    for (EventHandler *wrapper in handlers) {
        @try {
            wrapper.handler(processedData);
            self.eventsHandled++;
            
            if (wrapper.once) {
                [onceHandlers addObject:wrapper];
            }
        } @catch (NSException *exception) {
            NSLog(@"[EventBus] Handler error for %@: %@", event, exception);
            [self emit:@"bus:error" data:@{
                @"event": event,
                @"error": exception.description,
                @"data": processedData ?: @{}
            }];
        }
    }
    
    // Remove once handlers
    for (EventHandler *wrapper in onceHandlers) {
        [self.listeners[event] removeObject:wrapper];
    }
}

- (void)use:(BOOL (^)(NSString *, NSDictionary *))middleware {
    [self.middleware addObject:middleware];
}

- (NSDictionary *)getMetrics {
    return @{
        @"eventsEmitted": @(self.eventsEmitted),
        @"eventsHandled": @(self.eventsHandled)
    };
}

@end